File: ui/src/lib/api/permissions.ts
import { apiRequest } from './_base';

export type PermissionLevel = 'viewer' | 'user' | 'admin';

export interface ComponentDescriptor {
  key: string;
  label: string;
  category: 'core_financial' | 'secondary' | 'plugins' | 'admin' | string;
  description: string;
}

export interface ComponentCatalog {
  components: ComponentDescriptor[];
  levels: PermissionLevel[];
}

export interface EffectivePermission {
  component: string;
  role_level: PermissionLevel;
  grant_level: PermissionLevel | null;
  effective_level: PermissionLevel;
}

export interface UserPermissions {
  user_id: number;
  role: PermissionLevel;
  is_superuser: boolean;
  permissions: EffectivePermission[];
}

export interface PermissionAuditEntry {
  id: number;
  user_id: number;
  actor_user_id: number | null;
  actor_email: string | null;
  component: string;
  action: 'grant' | 'update' | 'revoke';
  previous_level: PermissionLevel | null;
  new_level: PermissionLevel | null;
  created_at: string;
}

export interface PermissionAuditResponse {
  entries: PermissionAuditEntry[];
}

const LEVEL_RANK: Record<PermissionLevel, number> = {
  viewer: 1,
  user: 2,
  admin: 3,
};

/** True iff `actual` is at least `required` in the viewer < user < admin order. */
export function hasLevel(actual: PermissionLevel, required: PermissionLevel): boolean {
  return LEVEL_RANK[actual] >= LEVEL_RANK[required];
}

export function getComponentCatalog(): Promise<ComponentCatalog> {
  return apiRequest<ComponentCatalog>('/permissions/components');
}

export function getMyPermissions(): Promise<UserPermissions> {
  return apiRequest<UserPermissions>('/permissions/me');
}

export function getUserPermissions(userId: number): Promise<UserPermissions> {
  return apiRequest<UserPermissions>(`/permissions/users/${userId}`);
}

export function setUserPermission(
  userId: number,
  component: string,
  level: PermissionLevel,
): Promise<EffectivePermission> {
  return apiRequest<EffectivePermission>(
    `/permissions/users/${userId}/components/${component}`,
    { method: 'PUT', body: JSON.stringify({ level }) },
  );
}

export function clearUserPermission(
  userId: number,
  component: string,
): Promise<EffectivePermission> {
  return apiRequest<EffectivePermission>(
    `/permissions/users/${userId}/components/${component}`,
    { method: 'DELETE' },
  );
}

export function getUserPermissionAudit(
  userId: number,
  limit = 100,
): Promise<PermissionAuditResponse> {
  return apiRequest<PermissionAuditResponse>(
    `/permissions/users/${userId}/audit?limit=${limit}`,
  );
}

File: api/core/schemas/permissions.py
"""Pydantic schemas for the per-user, per-component permission API."""

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

# Mirrors core.constants.components.PERMISSION_LEVELS.
PermissionLevel = Literal["viewer", "user", "admin"]


class ComponentDescriptor(BaseModel):
    key: str
    label: str
    category: str
    description: str


class ComponentCatalogResponse(BaseModel):
    components: List[ComponentDescriptor]
    levels: List[PermissionLevel]


class EffectivePermissionResponse(BaseModel):
    component: str
    role_level: PermissionLevel
    grant_level: Optional[PermissionLevel] = None
    effective_level: PermissionLevel


class UserPermissionsResponse(BaseModel):
    user_id: int
    role: PermissionLevel
    is_superuser: bool
    permissions: List[EffectivePermissionResponse]


class SetPermissionRequest(BaseModel):
    level: PermissionLevel = Field(
        ...,
        description="Permission level. Grants only restrict; they cannot elevate above the user's role.",
    )


class PermissionAuditEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    actor_user_id: Optional[int]
    actor_email: Optional[str]
    component: str
    action: Literal["grant", "update", "revoke"]
    previous_level: Optional[PermissionLevel]
    new_level: Optional[PermissionLevel]
    created_at: datetime


class PermissionAuditResponse(BaseModel):
    entries: List[PermissionAuditEntry]
